File: frontend/src/types/task.ts
/**
 * Типы данных для Task v1 и Execution v1
 * Recording Life System - Intent (намерение) + Fact (фиксация)
 */

// ============================================================================
// TASK V1 - INTENT (Намерение)
// ============================================================================

// Типы задач
export type TaskType = 'Task' | 'Bug' | 'Chore';

// Типы статуса (status = Done означает что ожидание больше не актуально!)
export type TaskStatus = 'Backlog' | 'Active' | 'Done' | 'Archived';

// Типы приоритета
export type Priority = 'Low' | 'Medium' | 'High';

// Типы места выполнения
export type Location = 'Home' | 'Office' | 'Anywhere';

// Типы connectivity
export type Connectivity = 'Online' | 'Offline';

// Типы рутины
export type RoutineType = 'Routine' | 'Ad-hoc';

/**
 * Контекст выполнения задачи
 */
export interface TaskContext {
  location?: Location;
  tools_required?: string[];
  connectivity?: Connectivity;
}

/**
 * Информация о повторяемости задачи
 */
export interface TaskRecurrence {
  is_repeatable: boolean;
  routine_type?: RoutineType;
  recurrence_rule?: string;  // Daily, Weekly, Cron expression
}

/**
 * Task v1 - Декларация намерения (Intent)
 *
 * Task отвечает на вопросы:
 * - Что это за ожидание?
 * - Насколько оно важно?
 * - Повторяется ли оно?
 * - В каком контексте его можно выполнять?
 */
export interface Task {
  // Идентификация и описание
  id: number;
  title: string;
  description?: string;
  type: TaskType;

  // Управление жизненным циклом
  status: TaskStatus;
  priority: Priority;
  created_at: string;
  updated_at: string;

  // Планирование (минимум)
  due_date?: string;

  // Контекст выполнения
  context?: TaskContext;

  // Рутинность и повторяемость
  recurrence?: TaskRecurrence;
}

/**
 * Схема для создания Task
 */
export interface TaskCreate {
  title: string;
  description?: string;
  type?: TaskType;
  status?: TaskStatus;
  priority?: Priority;
  due_date?: string;
  context?: TaskContext;
  recurrence?: TaskRecurrence;
}

/**
 * Схема для обновления Task
 */
export interface TaskUpdate {
  title?: string;
  description?: string;
  type?: TaskType;
  status?: TaskStatus;
  priority?: Priority;
  due_date?: string;
  context?: TaskContext;
  recurrence?: TaskRecurrence;
}

// ============================================================================
// EXECUTION V1 - FACT (Фиксация факта)
// ============================================================================

// Статусы выполнения
export type ExecutionStatus = 'completed' | 'failed' | 'skipped' | 'partial';

// Способ записи
export type RecordedBy = 'manual' | 'auto';

/**
 * Execution v1 - Фиксация факта выполнения
 *
 * Execution отвечает на вопросы:
 * - Что реально произошло?
 * - Когда?
 * - Чем закончилось?
 * - Какие данные я хочу сохранить по факту?
 */
export interface Execution {
  // Связь
  id: number;
  task_id: number;

  // Результат попытки
  status: ExecutionStatus;

  // Временные параметры
  started_at?: string;
  ended_at?: string;
  duration?: number;  // в секундах

  // Фиксация данных (черный ящик)
  payload?: Record<string, any>;
  note?: string;

  // Метаданные
  recorded_by: RecordedBy;
  created_at: string;
}

/**
 * Схема для создания Execution
 */
export interface ExecutionCreate {
  task_id: number;
  status: ExecutionStatus;
  started_at?: string;
  ended_at?: string;
  duration?: number;
  payload?: Record<string, any>;
  note?: string;
  recorded_by?: RecordedBy;
}

// ============================================================================
// WEBSOCKET MESSAGES
// ============================================================================

export interface WebSocketMessage {
  type: 'task_created' | 'task_updated' | 'task_deleted' | 'execution_created';
  task?: Task;
  execution?: Execution;
  task_id?: number;
  session_id?: string;
}

File: backend/app/schemas/task.py
"""
Pydantic схемы для валидации данных Task v1
"""
from pydantic import BaseModel, Field
from datetime import datetime, date
from typing import Optional, Dict, Any, List


# Вложенные схемы для context и recurrence
class TaskContext(BaseModel):
    """Контекст выполнения задачи"""
    location: Optional[str] = Field(None, description="Место выполнения: Home | Office | Anywhere")
    tools_required: Optional[List[str]] = Field(None, description="Необходимые инструменты")
    connectivity: Optional[str] = Field(None, description="Требование к сети: Online | Offline")


class TaskRecurrence(BaseModel):
    """Информация о повторяемости задачи"""
    is_repeatable: bool = Field(default=False, description="Повторяемая ли задача")
    routine_type: Optional[str] = Field(None, description="Тип рутины: Routine | Ad-hoc")
    recurrence_rule: Optional[str] = Field(None, description="Правило повторения (Daily, Weekly, Cron)")


class TaskBase(BaseModel):
    """
    Базовая схема Task v1 - декларация намерения (Intent)

    Task отвечает на вопросы:
    - Что это за ожидание?
    - Насколько оно важно?
    - Повторяется ли оно?
    - В каком контексте его можно выполнять?
    """
    # Идентификация и описание
    title: str = Field(..., min_length=1, max_length=200, description="Заголовок задачи")
    description: Optional[str] = Field(None, description="Описание задачи")
    type: str = Field(default="Task", description="Тип: Task | Bug | Chore")

    # Управление жизненным циклом
    status: str = Field(default="Backlog", description="Статус: Backlog | Active | Done | Archived")
    priority: str = Field(default="Medium", description="Приоритет: Low | Medium | High")

    # Планирование (минимум)
    due_date: Optional[date] = Field(None, description="Дедлайн (опционально)")

    # Контекст выполнения
    context: Optional[TaskContext] = Field(None, description="Контекст выполнения")

    # Рутинность и повторяемость
    recurrence: Optional[TaskRecurrence] = Field(None, description="Информация о повторяемости")


class TaskCreate(TaskBase):
    """Схема для создания задачи Task v1"""
    pass


class TaskUpdate(BaseModel):
    """Схема для обновления задачи Task v1 (все поля опциональны)"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    type: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    due_date: Optional[date] = None
    context: Optional[TaskContext] = None
    recurrence: Optional[TaskRecurrence] = None


class TaskResponse(TaskBase):
    """Схема для ответа с задачей Task v1"""
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True  # Для работы с ORM моделями
